modules/metastorage/src/main/java/org/apache/ignite/metastorage/internal/watch/WatchAggregator.java

```diff
@@ -0,0 +1,254 @@
+
+    /** */
+    private long idCntr;
+
+    /** */
+    public long add(Key key, long revision, WatchListener lsnr) {
+        var watch = new Watch(new ExactCriterion(key), revision, lsnr);
+        var id = idCntr++;
+        watches.put(id, watch);
+        return id;
+    }
+
+    /** */
+    public long addPrefix(Key key, long revision, WatchListener lsnr) {
+        var watch = new Watch(new PrefixCriterion(key), revision, lsnr);
+        var id = idCntr++;
+        watches.put(id, watch);
+        return id;
+    }
+
+    /** */
+    public long add(Collection<Key> keys, long revision, WatchListener lsnr) {
+        var watch = new Watch(new CollectionCriterion(keys), revision, lsnr);
+        var id = idCntr++;
+        watches.put(id, watch);
+        return id;
+    }
+
+    /** */
+    public long add(Key from, Key to, long revision, WatchListener lsnr) {
+        var watch = new Watch(new RangeCriterion(from, to), revision, lsnr);
+        var id = idCntr++;
+        watches.put(id, watch);
+        return id;
+    }
+
+    /** */
+    public void cancel(long id) {
+        watches.remove(id);
+    }
+
+    /** */
+    // TODO: We can do it better than infer range always
+    public IgniteBiTuple<Key, Key> inferGeneralCriteria() {
+        // TODO: Remove this assert with normal processing of empty watch
+        assert !watches.isEmpty();
+
+        return new IgniteBiTuple<>(
+            watches.values().stream().map(w -> w.keyCriterion.toRange().getKey()).min(Key::compareTo).get(),
+            watches.values().stream().map(w -> w.keyCriterion.toRange().getValue()).max(Key::compareTo).get()
+        );
+    }
+
+    /** */
+    public long inferGeneralRevision() {
+        // TODO: Remove this assert with normal processing of empty watch
+        assert !watches.isEmpty();
+
+        return Collections.min(watches.values().stream().map(w -> w.revision).collect(Collectors.toList()));
+    }
+
+    /** */
+    public WatchListener watchListener() {
+        // TODO: Remove this assert with normal processing of empty watch
+        assert !watches.isEmpty();
+
+        return new WatchListener() {
+
+            @Override public boolean onUpdate(@NotNull Iterable<WatchEvent> evts) {
+                Watch watch;
+                var watchIt = watches.values().iterator();
+                while (watchIt.hasNext()) {
+                    watch = watchIt.next();
+                    var filteredEvts = new ArrayList<WatchEvent>();
+
+                    for (WatchEvent evt: evts) {
+                       if (watch.keyCriterion.contains(evt.oldEntry().key()) &&
+                           (watch.revision < 0 || evt.oldEntry().revision() >= watch.revision))
+                           filteredEvts.add(evt);
+                    }
+
+                    if (!filteredEvts.isEmpty()) {
+                        if (!watch.lsnr.onUpdate(filteredEvts))
+                            watchIt.remove();
```

Comment:
Despite the fact that in general it'll work properly because of filtering, we probably should also reconfigure and redeploy big watch in some cases. For example if removed watch produced big-watch-range-from/to key, or if there are no more watches left.

Reply:
Will be addressed under https://issues.apache.org/jira/browse/IGNITE-14604